File: ios/Sources/ACCore/HostConfig/TeamsHostConfig.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import Foundation

/// Pre-configured Microsoft Teams host config loaded from bundled JSON files.
/// JSON source of truth lives in shared/host-configs/ios/.
public class TeamsHostConfig {
    private static var lightConfig: HostConfig?
    private static var darkConfig: HostConfig?

    /// Returns the default (light) Teams host config.
    public static func create() -> HostConfig { createLight() }

    /// Returns the light Teams host config, cached after first load.
    public static func createLight() -> HostConfig {
        if let cached = lightConfig { return cached }
        let config = loadBundledConfig("microsoft-teams-light")
        lightConfig = config
        return config
    }

    /// Returns the dark Teams host config, cached after first load.
    public static func createDark() -> HostConfig {
        if let cached = darkConfig { return cached }
        let config = loadBundledConfig("microsoft-teams-dark")
        darkConfig = config
        return config
    }

    private static func loadBundledConfig(_ name: String) -> HostConfig {
        guard let url = Bundle.module.url(forResource: name, withExtension: "json", subdirectory: "Resources") else {
            print("TeamsHostConfig: Missing bundled config \(name).json, using defaults")
            return HostConfig()
        }
        do {
            let data = try Data(contentsOf: url)
            return try HostConfigParser.parse(data)
        } catch {
            print("TeamsHostConfig: Failed to parse \(name).json — \(error)")
            return HostConfig()
        }
    }
}

File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/hostconfig/TeamsHostConfig.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.core.hostconfig

/**
 * Pre-configured HostConfig for Microsoft Teams loaded from bundled JSON files.
 * JSON source of truth lives in shared/host-configs/android/.
 *
 * Light and dark themes are loaded lazily and cached as singletons.
 */
object TeamsHostConfig {

    /** Creates the standard Teams light-theme HostConfig (cached singleton). */
    fun create(): HostConfig = lightInstance

    /** Teams light theme — white card surface (cached singleton). */
    fun createLight(): HostConfig = lightInstance

    /** Teams dark theme — dark card surface (cached singleton). */
    fun createDark(): HostConfig = darkInstance

    private val lightInstance: HostConfig by lazy {
        loadConfig("host-configs/microsoft-teams-light.json")
    }

    private val darkInstance: HostConfig by lazy {
        loadConfig("host-configs/microsoft-teams-dark.json")
    }

    private fun loadConfig(resourcePath: String): HostConfig {
        return try {
            val json = this::class.java.classLoader
                ?.getResourceAsStream(resourcePath)
                ?.bufferedReader()
                ?.readText()
            if (json != null) {
                HostConfigParser.parse(json)
            } else {
                println("TeamsHostConfig: Missing resource $resourcePath, using defaults")
                HostConfig()
            }
        } catch (e: Exception) {
            println("TeamsHostConfig: Failed to parse $resourcePath — ${e.message}")
            HostConfig()
        }
    }
}
